### PR TITLE
feat: crda cli: CLI Summary

### PR DESCRIPTION
**Pull request by deepak1725**
CRDA CLI: Summary for CLI Output

Example CLI output:
```
{"total_scanned_dependencies":0,
"direct_dependencies_with_known_vulnerabilites":1,
"direct_dependencies_with_synk_advisories":2,
"vulnerabilities_per_severity":{
     "low":2,
     "medium":1,
     "high":0,
     "critical":0
}
}
```

Cmd:
```
./crda analyse /
    --file /home/deepshar/projects/go/src/cli-tools/go.mod         # Absolute Path of Manifest File
    --debug # Optional
```

### CODE REVIEW

File: analyses/driver/matcher.go
Comment (arajkumar) [inline analyses/driver/matcher.go:15]: Create a separate package for the summay, all the types related to summary must go into the new package

File: cmd/analyse.go
Comment (arajkumar) [inline cmd/analyse.go:63]: I feel it is better to have a flag for the concise summary.

e.g. --summary

File: analyses/driver/matcher.go
Comment (arajkumar) [inline analyses/driver/matcher.go:31]: Does it stores counts?

File: analyses/driver/matcher.go
Comment (arajkumar) [inline analyses/driver/matcher.go:62]: You can ignore fields those are not used. Shall be added whenever the need arises instead of having a interface type.
```suggestion
	Dependencies           []interface{}          `json:"dependencies"`
	Ecosystem              string                 `json:"ecosystem"`
	LatestVersion          string                 `json:"latest_version"`
	Licenses               []interface{}          `json:"licenses"`
	Name                   string                 `json:"name"`
	PrivateVulnerabilities []VulnerabilitiesType  `json:"private_vulnerabilities"`
	PublicVulnerabilities  []VulnerabilitiesType  `json:"public_vulnerabilities"`
	RecommendedVersion     string                 `json:"recommended_version"`
	URL                    url.URL                `json:"url"`
	Version                string                 `json:"version"`
```

File: analyses/driver/matcher.go
Comment (arajkumar) [inline analyses/driver/matcher.go:66]: Irrelevent, but the typename seem to be weired.
```suggestion
type Response struct {
```

**Review by arajkumar — CHANGES_REQUESTED** [review]

File: analyses/driver/matcher.go
Comment (deepak1725) [reply analyses/driver/matcher.go:31]: Yes

File: analyses/driver/matcher.go
Comment (deepak1725) [reply analyses/driver/matcher.go:66]: Response can be `PostResponse` and `GetResponse`.. `Response` would be ambiguous to the reader

File: cmd/analyse.go
Comment (deepak1725) [reply cmd/analyse.go:63]: No, Summary is default behaviour as per ask. No flags
`--verbose` aka Detailed report is with flag.

File: analyses/driver/matcher.go
Comment (deepak1725) [reply analyses/driver/matcher.go]: Ack.

File: analyses/driver/matcher.go
Comment (deepak1725) [reply analyses/driver/matcher.go]: Ack.

File: analyses/stackanalyses/controller.go
Comment (arajkumar) [inline analyses/stackanalyses/controller.go:202]: This code should go into summary.

Basically the summary package should expose a function which takes SA result as an argument and returns a Summary struct.

Then you can call a function to print the Summary struct as a json. Decoupling summary computation and display will be more flexible for changes.

**Review by arajkumar — CHANGES_REQUESTED** [review]

File: analyses/summary/types.go
Comment (deepak1725) [reply analyses/summary/types.go]: Probably I got you wrong. Moved above types to `driver` package

File: analyses/stackanalyses/controller.go
Comment (deepak1725) [reply analyses/stackanalyses/controller.go]: Ack. Decoupled.

File: analyses/stackanalyses/controller.go
Comment (arajkumar) [inline analyses/stackanalyses/controller.go:208]: For the SRP rule, it would be great if you can split the dump to console part into separate function. But I'm still non convinced to dump json on console without a separate switch.
CC: @prashbnair

File: cmd/analyse.go
Comment (arajkumar) [inline cmd/analyse.go:65]: Is this behaviour documented through `--help` switch?

File: analyses/driver/matcher.go
Comment (arajkumar) [inline analyses/driver/matcher.go:55]: I guess I had asked to remove the unused field (which are typed as generic interface). Is there a reason for keeping this?

File: analyses/summary/helpers.go
Comment (arajkumar) [inline analyses/summary/helpers.go:1]: Can you add a test to verify the summary ?

File: analyses/summary/types.go
Comment (arajkumar) [inline analyses/summary/types.go:12]: @prashbnair Need your input for the summary part.

**Review by arajkumar — CHANGES_REQUESTED** [review]

File: analyses/driver/matcher.go
Comment (deepak1725) [reply analyses/driver/matcher.go:55]: UnknownDependencies would be used by coming up `verbose` flag

File: cmd/analyse.go
Comment (deepak1725) [reply cmd/analyse.go:65]: Do you mean to include this text in `--help` summary text of `analyse cmd` .?

File: analyses/driver/matcher.go
Comment (deepak1725) [reply analyses/driver/matcher.go]: Ack.

File: cmd/analyse.go
Comment (deepak1725) [reply cmd/analyse.go:66]: Help text of `analyse cmd` updated

File: analyses/driver/matcher.go
Comment (deepak1725) [reply analyses/driver/matcher.go]: Removed all unused items from this struct

File: analyses/summary/helpers.go
Comment (deepak1725) [reply analyses/summary/helpers.go:1]: Ack. Added Unit Test

File: analyses/stackanalyses/controller.go
Comment (prashbnair) [reply analyses/stackanalyses/controller.go:208]: I dont think we should dump json on the console without a flag

File: analyses/summary/types.go
Comment (prashbnair) [reply analyses/summary/types.go:12]: This looks fine as per the epic but can you take a sample output, add it to the epic and as Parag?

File: analyses/stackanalyses/controller.go
Comment (deepak1725) [reply analyses/stackanalyses/controller.go:208]: not sure, why we need a flag/switch. AFAIK Flags are optional, I mean CLI should work with and without them. Now consider a scenario with flag..? Does Analyse command hold any value without it..?

Comment (prashbnair) [reply analyses/stackanalyses/controller.go:208]: JSON is not the most readable format, it should be printed out in a form that is human readable. If anybody want a specific format , it can be done by a flag.

Comment (deepak1725) [reply analyses/stackanalyses/controller.go]: Ack.

File: analyses/stackanalyses/controller.go
Comment (arajkumar) [inline analyses/stackanalyses/controller.go:235]: It is not a common practice to add \n in the front. It hinders readability too.
```suggestion
	fmt.Println("Powered by Snyk\n")
	fmt.Println("Summary Report for Analyses:")
```

File: analyses/stackanalyses/controller.go
Comment (arajkumar) [inline analyses/stackanalyses/controller.go:248]: Ditto

File: analyses/stackanalyses/controller.go
Comment (arajkumar) [inline analyses/stackanalyses/controller.go:203]: Move summary related functionalities to summary pkg

File: analyses/stackanalyses/controller.go
Comment (arajkumar) [inline analyses/stackanalyses/controller.go:214]:
```suggestion
	return out.TotalVulnerabilities > 0
```

File: analyses/summary/helper_test.go
Comment (arajkumar) [inline analyses/summary/helper_test.go:67]: Use `github.com/google/go-cmp/cmp` to compare two structs directly.
```suggestion
	if diff := cmp.Diff(want, got); diff != "" {
	        t.Errorf("Vuln mismatch (-want, +got):\n%s", diff)
	}
```

**Review by arajkumar — CHANGES_REQUESTED** [review]

File: analyses/stackanalyses/controller.go
Comment (deepak1725) [reply analyses/stackanalyses/controller.go]: Ack.

File: analyses/summary/helper_test.go
Comment (deepak1725) [reply analyses/summary/helper_test.go]: Ack.

File: analyses/stackanalyses/controller.go
Comment (deepak1725) [reply analyses/stackanalyses/controller.go]: Ack.

File: analyses/stackanalyses/controller.go
Comment (deepak1725) [reply analyses/stackanalyses/controller.go]: Ack.

File: analyses/stackanalyses/controller.go
Comment (deepak1725) [reply analyses/stackanalyses/controller.go]: Ack.

**Review by arajkumar — APPROVED** [review]
lgtm, wait for @prashbnair approval.

File: cmd/analyse.go
Comment (prashbnair) [inline cmd/analyse.go:22]: Provides detailed report of vulnerabilities.

Comment (prashbnair) [reply cmd/analyse.go:22]: I dont think you need to add CRDA platform.

**Review by prashbnair — APPROVED** [review]

File: cmd/analyse.go
Comment (deepak1725) [reply cmd/analyse.go]: Ack

**Pull request merged**
